src/DotLiquid/Context.cs

```diff
@@ -510,38 +511,12 @@ private object Variable(string markup, bool notifyNotFound)
                 {
                     // If object is a KeyValuePair, we treat it a bit differently - we might be rendering
                     // an included template.
-                    if (@object is KeyValuePair<string, object> && ((KeyValuePair<string, object>)@object).Key == (string)part)
-                    {
-                        object res = ((KeyValuePair<string, object>)@object).Value;
-                        @object = Liquidize(res);
-                    }
-                    // If object is a hash- or array-like object we look for the
-                    // presence of the key and if its available we return it
-                    else if (IsHashOrArrayLikeObject(@object, part))
+                    object res = null;
+                    if (TryEvaluateHashOrArrayLikeObject(@object, part, out res))
                     {
                         // If its a proc we will replace the entry with the proc
-                        object res = LookupAndEvaluate(@object, part);
                         @object = Liquidize(res);
                     }
-                    // Some special cases. If the part wasn't in square brackets and
-                    // no key with the same name was found we interpret following calls
-                    // as commands and call them on the current object
-                    else if (!partResolved && (@object is IEnumerable) && (Template.NamingConvention.OperatorEquals(part as string, "size") || Template.NamingConvention.OperatorEquals(part as string, "first") || Template.NamingConvention.OperatorEquals(part as string, "last")))

```

Comment:
This logic is incorrectly removed. We should not compute first/last from square brackets. PartResolved is true if it matched square brackets. Also, the three other places using TryEvaluateHashOrArrayLikeObject now are computing first last when previously did not. This is a behavior change that seems unnecessary.

Reply:
As noted elsewhere this can be marked resolved

Reply:
See my other comment - this change is not consistent with reference Ruby liquid.

Reply:
Changes reverted